rootfs/etc/nginx/lua/balancer/leastconn.lua

```diff
@@ -0,0 +1,108 @@
+local util = require("util")
+local split = require("util.split")
+require("resty.core")
+
+local ngx = ngx
+local ipairs = ipairs
+local setmetatable = setmetatable
+local string_format = string.format
+local ngx_log = ngx.log
+local DEBUG = ngx.DEBUG
+
+
+local _M = { name = "leastconn" }
+
+function _M.new(self, backend)
+    local o = {
+        peers = backend.endpoints
+    }
+    setmetatable(o, self)
+    self.__index = self
+    return o
+end
+
+function _M.is_affinitized()
+    return false
+end
+
+
+local function get_upstream_name(upstream)
+    return upstream.address .. ":" .. upstream.port
+end
+
+
+function _M.balance(self)
+    local peers = self.peers
+    local endpoint = peers[1]
+    local endpoints = ngx.shared.balancer_leastconn
+    local feasible_endpoints = {}
+
+    if #peers ~= 1 then
+        local lowestconns = 2147483647
+        -- find the lowest connection count
+        for _, peer in pairs(peers) do
+            local peer_name = get_upstream_name(peer)
+            local peer_conn_count = endpoints:get(peer_name)
+            if peer_conn_count  == nil then
+                -- Peer has never been recorded as having connections - add it to the connection
+                -- tracking table and the list of feasible peers
+                endpoints:set(peer_name,0,0)
+                lowestconns = 0
+                feasible_endpoints[#feasible_endpoints+1] = peer
+            elseif peer_conn_count < lowestconns then
+                -- Peer has fewer connections than any other peer evaluated so far - add it as the
+                -- only feasible endpoint for now
+                feasible_endpoints = {peer}
+                lowestconns = peer_conn_count
+            elseif peer_conn_count == lowestconns then
+                -- Peer has equal fewest connections as other peers - add it to the list of
+                -- feasible peers
+                feasible_endpoints[#feasible_endpoints+1] = peer
+            end
+        end
+        ngx_log(DEBUG, "select from ", #feasible_endpoints, " feasible endpoints out of ", #peers)
+        endpoint = feasible_endpoints[math.random(1,#feasible_endpoints)]
+    end
+
+    local selected_endpoint = get_upstream_name(endpoint)
+    ngx_log(DEBUG, "selected endpoint ", selected_endpoint)
+
+    -- Update the endpoint connection count
+    endpoints:incr(selected_endpoint,1,1,0)
+
+    return selected_endpoint
+end
+
+function _M.after_balance(_)
+    local endpoints = ngx.shared.balancer_leastconn
+    local upstream = split.get_last_value(ngx.var.upstream_addr)
+
+    if util.is_blank(upstream) then
+        return
+    end
+    endpoints:incr(upstream,-1,0,0)
```

Comment:
I have a feeling there is a bug somewhere (most likely with removing/adding endpoints) which either calls this twice, or otherwise does this the "wrong" way;

<img width="1486" alt="image" src="https://github.com/kubernetes/ingress-nginx/assets/22740616/484fc603-a020-4828-8431-69ccd3955938">

These graphs show a massively divergent usage, all similar in increasing amount and such, which leads me to think that this either never reaches "0" after some TTL, or otherwise does not properly "reset" when it sees that it has gone below zero

Reply:
I know, that code is totally unnecessary, I've changed it around in my fork

(It still seems to exhibit an off-by-one error somewhere after indeterminate amount of time after an indeterminate event, which is why I haven't yet suggested it here)

Reply:
I've found the off-by-one error, and i've committed it here, if anyone's interested; https://github.com/ShadowJonathan/ingress-nginx/tree/add-least_connections-load-balancing